pkg/infrastructure/azure/azure.go

```diff
@@ -0,0 +1,356 @@
+	streamArch, err := stream.GetArchitecture(archName)
+	if err != nil {
+		return fmt.Errorf("failed to get rhcos architecture: %w", err)
+	}
+
+	azureDisk := streamArch.RHELCoreOSExtensions.AzureDisk
+	imageURL := azureDisk.URL
+
+	rawImageVersion := strings.ReplaceAll(azureDisk.Release, "-", "_")
+	imageVersion := rawImageVersion[:len(rawImageVersion)-6]
+
+	galleryName := fmt.Sprintf("gallery_%s", strings.ReplaceAll(in.InfraID, "-", "_"))
+	galleryImageName := in.InfraID
+	galleryImageVersionName := imageVersion
+	galleryGen2ImageName := fmt.Sprintf("%s-gen2", in.InfraID)
+	galleryGen2ImageVersionName := imageVersion
+
+	headResponse, err := http.Head(imageURL)
+	if err != nil {
+		return fmt.Errorf("failed HEAD request for image URL %s: %w", imageURL, err)
+	}
+
+	imageLength := headResponse.ContentLength
+	if imageLength%512 != 0 {
+		return fmt.Errorf("image length is not alisnged on a 512 byte boundary")
+	}
+
+	userTags := platform.UserTags
+	tags := make(map[string]*string, len(userTags)+1)
+	tags[fmt.Sprintf("kubernetes.io_cluster.%s", in.InfraID)] = ptr.To("owned")
+	for k, v := range userTags {
+		tags[k] = ptr.To(v)
+	}
+
+	tokenCredential := session.TokenCreds
+	storageURL := fmt.Sprintf("https://%s.blob.core.windows.net", storageAccountName)
+	blobURL := fmt.Sprintf("%s/%s/%s", storageURL, containerName, blobName)
+
+	// Create resource group
+	resourcesClientFactory, err := armresources.NewClientFactory(
+		subscriptionID,
+		tokenCredential,
+		&arm.ClientOptions{
+			ClientOptions: policy.ClientOptions{
+				Cloud: cloudConfiguration,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get azure resource groups factory: %w", err)
+	}
+	resourceGroupsClient := resourcesClientFactory.NewResourceGroupsClient()
+	resourceGroup, err := resourceGroupsClient.CreateOrUpdate(
+		ctx,
+		resourceGroupName,
+		armresources.ResourceGroup{
+			Location:  ptr.To(platform.Region),
+			ManagedBy: nil,
+			Tags:      tags,
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("error creating resource group %s: %w", resourceGroupName, err)
+	}
+	logrus.Debugf("ResourceGroup.ID=%s", *resourceGroup.ID)
+
+	// Create user assigned identity
+	userAssignedIdentityName := fmt.Sprintf("%s-identity", in.InfraID)
+	armmsiClientFactory, err := armmsi.NewClientFactory(subscriptionID, tokenCredential, nil)
+	if err != nil {
+		log.Fatalf("failed to create armmsi client: %v", err)
+	}
+	userAssignedIdentity, err := armmsiClientFactory.NewUserAssignedIdentitiesClient().CreateOrUpdate(
+		ctx,
+		resourceGroupName,
+		userAssignedIdentityName,
+		armmsi.Identity{
+			Location: ptr.To(platform.Region),
+			Tags:     tags,
+		},
+		nil,
+	)
+	if err != nil {
+		log.Fatalf("failed to create user assigned identity %s: %v", userAssignedIdentityName, err)
+	}
+	principalID := *userAssignedIdentity.Properties.PrincipalID
+
+	logrus.Debugf("UserAssignedIdentity.ID=%s", *userAssignedIdentity.ID)
+	logrus.Debugf("PrinciapalID=%s", principalID)
```

Comment:
This confused me at first, but these look like the identities we will use for the VMs :+1: